File: frontend/src/types/investigation.ts
export interface GraphNode {
  id: string;
  label: string;
  properties: Record<string, any>;
  x?: number;
  y?: number;
  vx?: number;
  vy?: number;
}

export interface GraphEdge {
  source: string;
  target: string;
  relation: string;
  properties: Record<string, any>;
}

export interface Citation {
  claim: string;
  evidence: string[];
}

export interface RelatedCrime {
  crime_id: string;
  record_id?: string;
  category?: string;
  connection_type: string;
  confidence_type?: "explicit" | "derived" | string;
  confidence?: number;
  similarity?: number;
}

export interface InvestigationResult {
  answer: string;
  structured_sections: {
    summary: string;
    connections: string;
    evidence: string;
    uncertainty: string;
  };
  citations: Citation[];
  evidence: Record<string, any>;
  graph_paths: GraphEdge[];
  graph_nodes: GraphNode[];
  related_crimes: RelatedCrime[];
  confidence: number;
  meta: Record<string, any>;
}

export interface GraphStats {
  status: string;
  is_fallback: boolean;
  crime_nodes: number;
  location_nodes: number;
  vehicle_nodes: number;
  weapon_nodes: number;
  mo_nodes: number;
  person_nodes: number;
  organization_nodes: number;
  cluster_nodes: number;
  pattern_nodes: number;
  total_nodes: number;
  total_relationships: number;
  explicit_relationships: number;
  derived_relationships: number;
  last_sync?: string | null;
}

export interface KeyIndividual {
  person_id: string;
  canonical_name: string;
  display_name: string;
  aliases: string[];
  source_provenance?: string;
  metrics: {
    degree_centrality: number;
    betweenness_centrality: number;
    pagerank: number;
    raw_degree: number;
  };
  network_breakdown: {
    connected_crimes: number;
    connected_people: number;
    connected_phones: number;
    connected_vehicles: number;
    connected_organizations: number;
    connected_locations: number;
  };
  structural_explanation?: string;
}

export interface ScopeOption {
  scope_type: string;
  scope_id: string;
  display_label: string;
  detail?: string;
}

export interface PersonProfile {
  person_id: string;
  canonical_name: string;
  aliases: string[];
  source_provenance: string;
  confidence: number;
  metrics?: {
    degree_centrality: number;
    betweenness_centrality: number;
    pagerank: number;
    raw_degree: number;
  };
  network_breakdown?: {
    connected_crimes: number;
    connected_people: number;
    connected_phones: number;
    connected_vehicles: number;
    connected_organizations: number;
    connected_locations: number;
  };
  structural_explanation?: string;
  crimes: Array<{
    crime_id: string;
    record_id?: string;
    category?: string;
    occurred_at?: string;
    role: string;
    relationship_type: string;
    evidence_excerpt?: string;
  }>;
  phones: Array<{
    phone_id: string;
    normalized_number: string;
    carrier?: string;
    role: string;
    confidence: number;
    metrics?: any;
  }>;
  telecom_summary: {
    total_phones: number;
    total_calls: number;
    total_airtime_seconds: number;
    top_contacts: any[];
  };
  subgraph?: {
    nodes: GraphNode[];
    edges: GraphEdge[];
  };
}

export type ReviewStatus = "PENDING" | "UNDER_REVIEW" | "VALIDATED" | "REJECTED" | "MODIFIED";

export interface ReviewHistoryItem {
  id: string;
  review_id: string;
  action: string;
  from_status?: string | null;
  to_status: string;
  reviewer_id: string;
  reviewer_display_name?: string | null;
  note?: string | null;
  reason?: string | null;
  metadata_snapshot?: Record<string, any>;
  created_at: string;
}

export interface ReviewItem {
  id: string;
  relationship_ref: string;
  source_entity_type: string;
  source_entity_id: string;
  target_entity_type: string;
  target_entity_id: string;
  relationship_type: string;
  original_relationship_type: string;
  final_relationship_type?: string | null;
  original_confidence: number;
  provenance: string;
  discovery_method?: string | null;
  status: ReviewStatus;
  reviewer_id?: string | null;
  reviewer_display_name?: string | null;
  investigator_note?: string | null;
  rejection_reason?: string | null;
  evidence_count: number;
  version: number;
  created_at: string;
  reviewed_at?: string | null;
  updated_at: string;
}

export interface ReviewDetail extends ReviewItem {
  evidence_snapshot: Array<{
    type?: string;
    id?: string;
    number?: string;
    call_id?: string;
    call_count?: number;
    total_duration?: number;
    excerpt?: string;
    note?: string;
    [key: string]: any;
  }>;
  history_entries: ReviewHistoryItem[];
}

export interface ReviewStats {
  total: number;
  pending: number;
  under_review: number;
  validated: number;
  rejected: number;
  modified: number;
}

// Phase 7: Domain AI Investigation Agent Types
export interface AgentNormalizedEvidence {
  evidence_id: string;
  evidence_type: string;
  source_system: string;
  source_record_id?: string | null;
  source_entity?: string | null;
  target_entity?: string | null;
  relationship?: string | null;
  summary: string;
  confidence: number;
  validation_status: string;
  citation: string;
  tool_used: string;
  metadata?: Record<string, any>;
  timestamp?: string | null;
}

export interface AgentFinding {
  title: string;
  details: string;
  supporting_evidence_ids: string[];
  validation_status: string;
  confidence?: number | null;
}

export interface AgentKeyIndividual {
  person_id: string;
  display_name: string;
  degree_centrality: number;
  betweenness_centrality: number;
  pagerank: number;
  connected_crimes: string[];
  connected_phones: string[];
  validation_status: string;
}

export interface AgentToolTraceItem {
  step: number;
  tool: string;
  purpose: string;
  status: "SUCCESS" | "FAILED" | "SKIPPED";
  duration_ms: number;
  result_count: number;
  error?: string | null;
}

export interface AgentInvestigationResponse {
  investigation_id: string;
  question: string;
  status: string;
  summary: string;
  findings: AgentFinding[];
  key_individuals: AgentKeyIndividual[];
  evidence: AgentNormalizedEvidence[];
  uncertainties: string[];
  limitations: string[];
  tool_trace: AgentToolTraceItem[];
  citations: string[];
  graph_data?: {
    nodes: Array<{ id: string; label: string; type: string; status: string }>;
    links: Array<{ source: string; target: string; relationship: string; status: string }>;
  } | null;
  iterations_count: number;
  total_duration_ms: number;
}

export interface AgentToolDefinition {
  name: string;
  description: string;
  input_schema: Record<string, any>;
  output_schema: Record<string, any>;
  permission: string;
  timeout_seconds: number;
  max_results: number;
  provenance_type: string;
}

export interface AgentExampleItem {
  title: string;
  scenario: string;
  question: string;
}

// ---------------------------------------------------------------------------
// Phase 8: Unified Investigation Command Center Types
// ---------------------------------------------------------------------------
export interface ExtractedEntitySummary {
  id: string;
  type: "Person" | "Phone" | "Vehicle" | "Weapon" | "Organization" | "Location" | string;
  label: string;
  role?: string | null;
  confidence: number;
  validation_status: string;
}

export interface ConnectedCaseSummary {
  crime_id: string;
  record_id: string;
  category: string;
  location_name: string;
  connection_type: string;
  confidence: number;
  evidence: string;
}

export interface KeyIndividualSummary {
  person_id: string;
  name: string;
  degree_centrality: number;
  betweenness_centrality: number;
  pagerank: number;
  structural_explanation?: string | null;
}

export interface CaseContextBundle {
  case_id: string;
  record_id: string;
  category: string;
  crime_type: string;
  description: string;
  location_name: string;
  latitude?: number | null;
  longitude?: number | null;
  occurred_at?: string | null;
  source: string;
  extracted_people: ExtractedEntitySummary[];
  extracted_phones: ExtractedEntitySummary[];
  extracted_vehicles: ExtractedEntitySummary[];
  extracted_weapons: ExtractedEntitySummary[];
  connected_cases: ConnectedCaseSummary[];
  key_individuals: KeyIndividualSummary[];
  validation_summary: {
    validated: number;
    under_review: number;
    rejected: number;
    modified: number;
  };
  total_evidence_count: number;
}

export interface TimelineEventItem {
  id: string;
  timestamp: string;
  event_type: "CRIME_INCIDENT" | "CDR_COMMUNICATION" | "INVESTIGATOR_DECISION" | "RELATIONSHIP_DISCOVERED" | "AGENT_FINDING" | string;
  source: string;
  title: string;
  description: string;
  entity_id?: string | null;
  entity_name?: string | null;
  entity_type?: string | null;
  validation_status?: string | null;
  citation?: string | null;
  metadata?: Record<string, any> | null;
}

export interface CaseTimelineData {
  case_id: string;
  events: TimelineEventItem[];
  total_events: number;
}

export interface SearchHitItem {
  id: string;
  entity_type: "CRIME" | "PERSON" | "PHONE" | "REVIEW";
  title: string;
  subtitle: string;
  category?: string | null;
  detail?: string | null;
  metadata?: Record<string, any> | null;
}

export interface GlobalSearchResult {
  query: string;
  crimes: SearchHitItem[];
  people: SearchHitItem[];
  phones: SearchHitItem[];
  reviews: SearchHitItem[];
  total_results: number;
}

export interface InvestigationReportData {
  case_id: string;
  record_id: string;
  title: string;
  generated_at: string;
  scope: Record<string, any>;
  executive_summary: string;
  connected_cases: ConnectedCaseSummary[];
  people_findings: Array<Record<string, any>>;
  telecom_findings: Array<Record<string, any>>;
  network_findings: Array<Record<string, any>>;
  timeline: TimelineEventItem[];
  geographic_findings: Array<Record<string, any>>;
  evidence_inventory: Array<Record<string, any>>;
  validation_breakdown: Record<string, number>;
  uncertainties: string[];
  methodology: string[];
  statutory_limitations: string[];
}

File: backend/app/schemas/investigation.py
"""
Phase 4 & Phase 8: Investigation Intelligence Pydantic Schemas
Provides data contracts for Graph RAG, Unified Case Context, Multi-Source Timeline,
Global Search, and Comprehensive Investigation Reports.
"""
from typing import Dict, Any, List, Optional
from datetime import datetime
from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Phase 4 Schemas: Grounded Graph RAG
# ---------------------------------------------------------------------------
class InvestigationQueryRequest(BaseModel):
    question: str = Field(..., min_length=2, max_length=1000, description="Natural-language crime investigation question")


class Citation(BaseModel):
    claim: str
    evidence: List[str] = Field(default_factory=list, description="Authoritative Crime IDs supporting this claim")


class RelatedCrime(BaseModel):
    crime_id: str
    record_id: Optional[str] = None
    category: Optional[str] = None
    connection_type: str
    confidence_type: Optional[str] = "derived"
    confidence: Optional[float] = None
    similarity: Optional[float] = None


class InvestigationQueryResponse(BaseModel):
    answer: str
    structured_sections: Dict[str, str] = Field(
        default_factory=dict,
        description="Structured sections: summary, connections, evidence, uncertainty"
    )
    citations: List[Citation] = Field(default_factory=list)
    evidence: Dict[str, Any] = Field(default_factory=dict)
    graph_paths: List[Dict[str, Any]] = Field(default_factory=list)
    graph_nodes: List[Dict[str, Any]] = Field(default_factory=list)
    related_crimes: List[RelatedCrime] = Field(default_factory=list)
    confidence: float = 0.85
    meta: Dict[str, Any] = Field(default_factory=dict)


class InvestigationExamplesResponse(BaseModel):
    examples: List[str]


# ---------------------------------------------------------------------------
# Phase 8 Schemas: Unified Investigation Command Center
# ---------------------------------------------------------------------------
class ExtractedEntitySummary(BaseModel):
    id: str
    type: str  # "Person", "Phone", "Vehicle", "Weapon", "Organization", "Location"
    label: str
    role: Optional[str] = None
    confidence: float = 1.0
    validation_status: str = "AI_DERIVED"


class ConnectedCaseSummary(BaseModel):
    crime_id: str
    record_id: str
    category: str
    location_name: str
    connection_type: str
    confidence: float
    evidence: str


class KeyIndividualSummary(BaseModel):
    person_id: str
    name: str
    degree_centrality: float
    betweenness_centrality: float
    pagerank: float
    structural_explanation: Optional[str] = None


class CaseContextResponse(BaseModel):
    case_id: str
    record_id: str
    category: str
    crime_type: str
    description: str
    location_name: str
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    occurred_at: Optional[str] = None
    source: str
    extracted_people: List[ExtractedEntitySummary] = Field(default_factory=list)
    extracted_phones: List[ExtractedEntitySummary] = Field(default_factory=list)
    extracted_vehicles: List[ExtractedEntitySummary] = Field(default_factory=list)
    extracted_weapons: List[ExtractedEntitySummary] = Field(default_factory=list)
    connected_cases: List[ConnectedCaseSummary] = Field(default_factory=list)
    key_individuals: List[KeyIndividualSummary] = Field(default_factory=list)
    validation_summary: Dict[str, int] = Field(
        default_factory=lambda: {"validated": 0, "under_review": 0, "rejected": 0, "modified": 0}
    )
    total_evidence_count: int = 0


class TimelineEvent(BaseModel):
    id: str
    timestamp: str  # ISO-8601 format
    event_type: str  # "CRIME_INCIDENT", "CDR_COMMUNICATION", "INVESTIGATOR_DECISION", "RELATIONSHIP_DISCOVERED", "AGENT_FINDING"
    source: str      # "POSTGRESQL", "CDR", "REVIEW_QUEUE", "AI_AGENT"
    title: str
    description: str
    entity_id: Optional[str] = None
    entity_name: Optional[str] = None
    entity_type: Optional[str] = None  # "Crime", "Person", "Phone"
    validation_status: Optional[str] = None  # "VALIDATED", "AI_DERIVED", "REJECTED", "MODIFIED"
    citation: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None


class CaseTimelineResponse(BaseModel):
    case_id: str
    events: List[TimelineEvent] = Field(default_factory=list)
    total_events: int = 0


class SearchItem(BaseModel):
    id: str
    entity_type: str  # "CRIME", "PERSON", "PHONE", "REVIEW"
    title: str
    subtitle: str
    category: Optional[str] = None
    detail: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None


class GlobalSearchResponse(BaseModel):
    query: str
    crimes: List[SearchItem] = Field(default_factory=list)
    people: List[SearchItem] = Field(default_factory=list)
    phones: List[SearchItem] = Field(default_factory=list)
    reviews: List[SearchItem] = Field(default_factory=list)
    total_results: int = 0


class InvestigationReportResponse(BaseModel):
    case_id: str
    record_id: str
    title: str
    generated_at: str
    scope: Dict[str, Any]
    executive_summary: str
    connected_cases: List[ConnectedCaseSummary] = Field(default_factory=list)
    people_findings: List[Dict[str, Any]] = Field(default_factory=list)
    telecom_findings: List[Dict[str, Any]] = Field(default_factory=list)
    network_findings: List[Dict[str, Any]] = Field(default_factory=list)
    timeline: List[TimelineEvent] = Field(default_factory=list)
    geographic_findings: List[Dict[str, Any]] = Field(default_factory=list)
    evidence_inventory: List[Dict[str, Any]] = Field(default_factory=list)
    validation_breakdown: Dict[str, int] = Field(default_factory=dict)
    uncertainties: List[str] = Field(default_factory=list)
    methodology: List[str] = Field(default_factory=list)
    statutory_limitations: List[str] = Field(default_factory=list)
